token-limiter-concurrent/src/Control/Concurrent/TokenLimiter/Concurrent.hs

```diff
@@ -92,16 +103,31 @@ tryDebit TokenLimiter {..} debit = modifyMVar tokenLimiterLastServiced $ \(lastS
       pure ((now, newCount), True)
     else pure ((lastServiced, countThen), False)
 
--- | Wait until the given number of tokens can be debited
+-- | Wait until the given number of tokens can be debited.
+--
+-- Note: debitor threads are serviced in FIFO order, so a request for a small
+-- (and currently satisfiable) number of tokens can still be delayed by a debit
+-- request for a larger amount of tokens.
 waitDebit :: TokenLimiter -> Word64 -> IO ()
-waitDebit TokenLimiter {..} debit = modifyMVar_ tokenLimiterLastServiced $ \(lastServiced, countThen) -> do
+waitDebit limiter debit = void (waitDebitObserve limiter debit)
+
+-- | Same as 'waitDebit', but returns the wait time due to rate limiting.
+--
+-- Note: only reports the time waited due to rate-limiting this specific action,
+-- not the wall-clock time waited (that might include waiting for previous
+-- actions limited by this rate limiter to finish).
+--
+-- Note: the wait time reported can be inflated due to scheduling inaccuracy.
+-- See https://gitlab.haskell.org/ghc/ghc/-/issues/16601.
+waitDebitObserve :: TokenLimiter -> Word64 -> IO (Maybe MonotonicDiffNanos)
+waitDebitObserve TokenLimiter {..} debit = modifyMVar tokenLimiterLastServiced $ \(lastServiced, countThen) -> do
```

Comment:
Oh now I see what you meant about the blocking.
I wonder if this could be improved.

Reply:
Not sure, I wouldn't stress about it - `canDebit` already documents that its information can get stale quickly, so even if one would report "yes, can" quickly without blocking, no guarantee it would hold true on subsequent `waitDebit`.

Rebasing the internals to work on STM (and well, actually exposing it, so canDebit / waitDebit would be an STM action) this token accounting would be easier, but I expect the timed wait and fairness bits would then be harder. unsafeIoToSTM could work with the timed wait, but the fairness might need a TQueue or such.

Reply:
`tryDebit` has the same problem though, which defeats its purpose a bit.
I'll think about this but it's not blocking this PR.